button-inc/emissions-elt-demo
# fix: nextauth unable to login user

Addresses #226. This fix corrects the issue of a user being unable to log in using the options presented by NextAuth. In service of that, GitHub configuration has been added. 

Thanks to @YaokunLin and @shon-button (and everyone else) for assists on this one!

## Changes

- Added configuration keys for GitHub (secrets added in GCP secrets page).
- Added missing .env variables to Docker/cloud build for database connection and NextAuth.
- Changed callback to a relative path.
- Added secrets to configuration to use nip.io to satisfy TLD requirements of GCP's auth.

## To test

- Navigate to [34.125.36.177.nip.io](http://34.125.36.177.nip.io/).
- Choose the Sign in with Google option.
  - You should eventually be redirected to the Climate Trax dashboard!
- _Optional:_ If your default email for GitHub is `@button.is`, open an incognito/private window and navigate to the site.
  - Sign in with GitHub.
  - As above, you should end up on the Climate Trax dashboard.

### Notes

#### GitHub auth

GitHub is currently authenticating against an OAuth key in my GitHub account. As this was used for testing (as Google Auth would not work initially due to the requirement below), and we don't necessarily need to have it as an option in the long term, this should be sufficient. If we [subsequently need to use GitHub OAuth](https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/authorizing-oauth-apps), I'll transfer ownership into the Button GitHub account (which would disable my editing access and make it tough to test). 

#### Nip.io

[Nip.io](https://nip.io/) is used as a DNS forwarder to satisfy the requirement to use a [TLD and not an IP when authenticating with Google auth](https://github.com/google/guava/wiki/InternetDomainNameExplained#public-suffixes-and-private-domains).

## Further info

Lots of these issues loop back to environment variables being exposed by Docker and having them stream down into the build from Google Secrets Manager. Remember to check your `.env`s and env variables!
